core/src/main/java/org/infinispan/cache/impl/StatsCollectingCache.java

```diff
@@ -0,0 +1,356 @@
+         statsCollector.recordStores(1, end - start);
+      }
+      return ret;
+   }
+
+   @Override
+   protected void putForExternalReadInternal(K key, V value, Metadata metadata, ByRef.Boolean isCreatedRef) {
+      boolean statisticsEnabled = statsCollector.getStatisticsEnabled();
+      long start = 0;
+      if (statisticsEnabled) {
+         start = timeService.time();
+      }
+      super.putForExternalReadInternal(key, value, metadata, isCreatedRef);
+      if (statisticsEnabled && isCreatedRef.get()) {
+         long end = timeService.time();
+         statsCollector.recordStores(1, end - start);
+      }
+   }
+
+   @Override
+   protected V putIfAbsentInternal(K key, V value, Metadata metadata) {
+      boolean statisticsEnabled = statsCollector.getStatisticsEnabled();
+      long start = 0;
+      if (statisticsEnabled) {
+         start = timeService.time();
+      }
+      V ret = super.putIfAbsentInternal(key, value, metadata);
+      if (statisticsEnabled && ret == null) {
+         long end = timeService.time();
+         statsCollector.recordStores(1, end - start);
+      }
+      return ret;
+   }
+
+   @Override
+   public V remove(Object key) {
+      boolean statisticsEnabled = statsCollector.getStatisticsEnabled();
+      long start = 0;
+      if (statisticsEnabled) {
+         start = timeService.time();
+      }
+      V ret = super.remove(key);
+      if (statisticsEnabled) {
+         long end = timeService.time();
+         if (ret != null) {
+            statsCollector.recordRemoveHits(1, end - start);
+         } else {
+            statsCollector.recordRemoveMisses(1);
+         }
+      }
+      return ret;
+   }
+
+   @Override
+   public boolean remove(Object key, Object value) {
+      boolean statisticsEnabled = statsCollector.getStatisticsEnabled();
+      long start = 0;
+      if (statisticsEnabled) {
+         start = timeService.time();
+      }
+      boolean removed = super.remove(key, value);
+      if (statisticsEnabled) {
+         long end = timeService.time();
+         if (removed) {
+            statsCollector.recordRemoveHits(1, end - start);
+         } else {
+            statsCollector.recordRemoveMisses(1);
+         }
+      }
+      return removed;
+   }
+
+   @Override
+   protected boolean replaceInternal(K key, V oldValue, V value, Metadata metadata) {
+      boolean statisticsEnabled = statsCollector.getStatisticsEnabled();
+      long start = 0;
+      if (statisticsEnabled) {
+         start = timeService.time();
+      }
+      boolean replaced = super.replaceInternal(key, oldValue, value, metadata);
+      if (statisticsEnabled && replaced) {
+         long end = timeService.time();
+         statsCollector.recordStores(1, end - start);
+      }
+      return replaced;
+   }
+
+   @Override
+   public CloseableIteratorSet<Entry<K, V>> entrySet() {
+      return new EntrySet() {
+         @Override
+         protected int addAllInternal(Collection<? extends Entry<K, V>> c) {
```

Comment:
I guess @wburns was a bit overzealous with the implementation :) The `Map.entrySet()` javadoc says `add()` isn't supported, so I think it would be best to remove it.